src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs

```diff
@@ -0,0 +1,415 @@
+                        // lowered decision dag, jump there to start.
+                        _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(wc)));
+                        break;
+                    case BoundDecision d:
+                        // If the first node is a decision rather than the code for the
+                        // lowered decision dag, jump there to start.
+                        _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(d)));
+                        break;
+                }
+
+                // Call LowerDecisionDagNode with each node and its following node in the generation order.
+                // However, some nodes can be emitted more efficiently as a switch dispatch (possibly out
+                // of order).
+                var loweredNodes = PooledHashSet<BoundDecisionDag>.GetInstance();
+                BoundDecisionDag previous = null;
+                foreach (BoundDecisionDag node in sortedNodes)
+                {
+                    if (node is BoundWhenClause w)
+                    {
+                        // Code for the when clause goes in a separate code section for the switch section.
+                        LowerWhenClause(w);
+                        continue;
+                    }
+
+                    // BoundDecision nodes do not get any code generated for them.
+                    if (node is BoundDecision || loweredNodes.Contains(node))
+                    {
+                        continue;
+                    }
+
+                    if (previous != null && !loweredNodes.Contains(previous))
+                    {
+                        // Lower the node "previous".
+                        if (!GenerateSwitchDispatch(previous, loweredNodes))
+                        {
+                            LowerDecisionDagNode(previous, node);
+                        }
+
+                        loweredNodes.Add(previous);
+                    }
+
+                    previous = node;
+                }
+
+                // Lower the final node
+                if (previous != null && !loweredNodes.Contains(previous))
+                {
+                    LowerDecisionDagNode(previous, null);
+                }
+
+                loweredNodes.Free();
+            }
+
+            /// <summary>
+            /// Generate a switch dispatch for a contiguous sequence of dag nodes if applicable.
+            /// Returns true if it was applicable.
+            /// </summary>
+            private bool GenerateSwitchDispatch(BoundDecisionDag node, HashSet<BoundDecisionDag> loweredNodes)
+            {
+                Debug.Assert(!loweredNodes.Contains(node));
+
+                // We only generate a switch dispatch if we have two or more value decisions in a row
+                if (!(node is BoundDecisionPoint firstDecisionPoint &&
+                     firstDecisionPoint.Decision is BoundNonNullValueDecision firstDecision &&
```

Comment:
There are a lot of `&&`s in this `if` condition, and in the `while` condition below. Could it be simplified (slightly) by extracting a helper method for `node is BoundDecisionPoint dp && dp.Decision is BoundNonNullValueDecision decision`? (That expression appears several times.)

Reply:
Perhaps not since it looks like the conditions also use the intermediate `BoundDecisionPoint` values as well. 

---
In reply to: [176305766](https://github.com/dotnet/roslyn/pull/25544#discussion_r176305766) [](ancestors = 176305766)